TypeScript section (#339)
vuejs/docs-next: src/guide/typescript-support.md

  setup(props) {
    const result = props.message.split('') // correct, 'message' is typed as a string
    const filtered = props.message.filter(p => p.value) // an error will be thrown: Property 'filter' does not exist on type 'string'
  }
})
```

### Typing `ref`s

Refs infer the type from the initial value:

```ts
import { defineComponent, ref } from 'vue'

const Component = defineComponent({
  setup() {
    const year = ref(2020)

    const result = year.value.split('') // => Property 'filter' does not exist on type 'number'
  }
})
```

Sometimes we may need to specify complex types for a ref's inner value. We can do that simply passing a generic argument when calling ref to override the default inference:

```ts
const year = ref<string | number>('2020') // year's type: Ref<string | number>

year.value = 2020 // ok!
```

::: tip Note
If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`.
:::

### Typing `reactive`

When typing a `reactive` property, we can use use interfaces:

```ts
import { defineComponent, reactive } from 'vue'

interface Book {
  title: string
  year?: number
}

export default defineComponent({
  name: 'HelloWorld',
  setup() {

>     const book: Book = reactive({ title: 'Vue 3 Guide' })

Reviewer: ```suggestion
    const book = reactive<Book>({ title: 'Vue 3 Guide' })
    // or
    const book: Book = reactive({ title: 'Vue 3 Guide' })
```

Just adding the other possible example

  }
})
```

Sometimes we may need to specify complex types for a ref's inner value. We can do that simply passing a generic argument when calling ref to override the default inference:

```ts
const year = ref<string | number>('2020') // year's type: Ref<string | number>

year.value = 2020 // ok!
```

::: tip Note
If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`.
:::

### Typing `reactive`

When typing a `reactive` property, we can use use interfaces:

```ts
import { defineComponent, reactive } from 'vue'

interface Book {
  title: string
  year?: number
}

export default defineComponent({
  name: 'HelloWorld',
  setup() {
    const book = reactive<Book>({ title: 'Vue 3 Guide' })
    // or
    const book: Book = reactive({ title: 'Vue 3 Guide' })
    // or
    const book = reactive({ title: 'Vue 3 Guide' }) as Book
  }
})
```

### Typing `computed`

Computed values will automatically infer the type from returned value

```ts
import { defineComponent, ref, computed } from 'vue'

export default defineComponent({
  name: 'HelloWorld',
  setup() {
    let count = ref(0)

    // read-only

>     const double = computed(() => count.value)

Reviewer 2: Compute a value without any change is IMHO a little too simply..

You can type your `data` properties using type assertions:

```js
const Component = defineComponent({
  data() {
    return {
      message: 'Hello!' as string,
      count: 0 as number
    }
  },
  mounted() {
    const result = this.count.split('') // => Property 'split' does not exist on type 'number'
  }
})
```

If you need a complex type, you can create an interface:

```ts
interface Book {
  title: string
  author: string
  year: number
}

const Component = defineComponent({
  data() {
    return {
      message: 'Hello!' as string,
      count: 0 as number,
      book: <Book>{
        title: 'Vue 3 Guide',
        author: 'Vue Team',
        year: 2020
      }
    }
  }
})
```

### Annotating Return Types

Because of the circular nature of Vue’s declaration files, TypeScript may have difficulties inferring the types of certain methods. For this reason, you may need to annotate the return type on methods like render and those in computed.

```ts
import { defineComponent } from 'vue'

const Component = defineComponent({
  data() {
    return {
      message: 'Hello!' as string
    }
  },
  methods: {
    // need annotation due to `this` in return type

>     greet(): string {

Reviewer: I don't think the annotation here is necessary, only `computed` is required, I added a dts check just to be sure 😄 
https://github.com/vuejs/vue-next/pull/1708

      return this.message + ' world'
    }
  },

>   computed: {

Reviewer: Should we add an example with `get` and `set`? 

just added to DTS tests: https://github.com/vuejs/vue-next/pull/1709

Author: Yup, I will add it to be consistent with Composition API part

    // need annotation
    greeting(): string {
      return this.message + '!'
    }
  }
})
```

### Annotating Props

> 

Reviewer: We should give some explaination:

"Vue does runtime validation on `props` to provide typescript we need to cast the constructor with `PropType`." - as usual better working advised 😄

```ts

> { defineComponent, PropType } from 'vue'

Reviewer: ```suggestion
import { defineComponent, PropType } from 'vue'
```

interface ComplexMessage {
  title: string
  okMessage: string
  cancelMessage: string
}
const Component = defineComponent({
  props: {
    name: String,
    success: { type: String },
    callback: {
      type: Function as PropType<() => void>
    },
    message: {
      type: Object as PropType<ComplexMessage>,
      required: true,
      validator(message: ComplexMessage) {
        return !!message.title
      }
    }
  }
})
```

If you find validator not getting type inference or member completion isn’t working, annotating the argument with the expected type may help address these problems.

## Using with Composition API

> On `setup()` function, you don't need ass a typing to `props` parameter as it will infer types from `props` component option.

Reviewer: ```suggestion
On `setup()` function, you don't need pass a typing to `props` parameter as it will infer types from `props` component option.
```

```ts
import { defineComponent } from 'vue'

const Component = defineComponent({
  props: {
    message: {
      type: String,
      required: true
    }
  },

  setup(props) {
    /* props will be typed as

>     (parameter) props: Readonly<{

Reviewer: No sure about providing the full type here, probably just showing the `props.message.split('')` is valid because is string

  setup(props) {
    const result = props.message.split('') // correct, 'message' is typed as a string
    const filtered = props.message.filter(p => p.value) // an error will be thrown: Property 'filter' does not exist on type 'string'
  }
})
```

### Typing `ref`s

To type a ref, we need to specify a type when calling `ref()` method:

```ts
import { defineComponent, ref } from 'vue'

const Component = defineComponent({
  setup() {

>     const year = ref<number>(2020)

Reviewer: ```ts
    const year = ref(2020) // equivalent to ref<number>(2020)
```
In this cases `ref` should infer the type

  }
})
```

> Sometimes we may need to specify complex types for a ref's inner value. We can do that succinctly by passing a generics argument when calling ref to override the default inference:

Reviewer: ```suggestion
Sometimes we may need to specify complex types for a ref's inner value. We can do that simply passing a generic argument when calling ref to override the default inference:
```

Simplifying wording

```ts
const year = ref<string | number>('2020') // year's type: Ref<string | number>

year.value = 2020 // ok!
```

> If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`:

Reviewer: Not sure about this paragraph, the following example `State` is not `unknown`, I think it will make it better by just having, when using with generic is recommended to cast ...

Author: Removed an example 👍🏻

```ts
function useState<State extends string>(initial: State) {
  const state = ref(initial) as Ref<State> // state.value -> State extends string
  return state
}
```

### Typing `reactive`

> Similarly to `ref`, we can define a `reactive` type on method call:

Reviewer: Probably we can mention that `reactive` should have the same type as the argument, only if there's unwrapped ref, that's when the types will be different

Author: @pikax if we assign this to `book`:

```ts
const book: Book = reactive({ title: "Vue 3 Guide", year: ref(2020) });
```

There is no error thrown. However, if we try to reassign the property, error is present:

```ts
const book: Book = reactive({ title: "Vue 3 Guide" });

book.year = ref(2020) // 'Ref<number>' is not assignable to type 'number'
```

Is there an unwrapping happening somewhere? 🤔 
```

Reviewer: Yes, there's unwrapping when you use `reactive`. 

Also when you assign a `ref` to a value in `reactive` it will `unwrap`, but unfortunate  we can't do it in typescript because it doesn't allow us to define a setter with a different type from getter https://github.com/microsoft/TypeScript/issues/2521

Author: Gotcha :( Should we leave this example as it is now or it's better to remove it?

Reviewer: I would change the example to: 
```ts
const book: Book = reactive({ title: "Vue 3 Guide", year: ref(20)});
```

Author: @pikax but it won't throw an error 🤔 

I don't mind changing it anyways 👍🏻

Reviewer: Do you want to throw error?

```ts
import { defineComponent, reactive } from 'vue'

interface Book {
  title: string
  year?: number
}

export default defineComponent({
  name: 'HelloWorld',
  setup() {
    const book = reactive<Book>({
      title: 'Vue 3 Guide'
    })
  }
})
```

### Typing `computed`

For computed, we could specify a type with getter only, or with getter and setter:

```ts
import { defineComponent, ref, rcomputed } from "vue";

export default defineComponent({
  name: "HelloWorld",
  setup() {
    let count = ref<number>(0);

>     // read-only, will have a type of Readonly<Ref<Readonly<number>>>

Reviewer: Although  technically correct, must users won't understand the type here.

>     const book = reactive<Book>({

Reviewer: Here we can actually show that the `reactive` can be casted to the `Book` because if the `Book` doesn't have any nested `ref`, it will be pretty much the same as `Book`.

In this case, this would be equivalent: 
```ts
const book: Book = reactive({ title: 'Vue 3 Guide' });
// or 
const book = reactive({ title: 'Vue 3 Guide' }) as Book;
```

Author: @pikax if we assign this to `book`:

```ts
const book: Book = reactive({ title: "Vue 3 Guide", year: ref(2020) });
```

There is no error thrown. However, if we try to reassign the property, error is present:

```ts
const book: Book = reactive({ title: "Vue 3 Guide" });

book.year = ref(2020) // 'Ref<number>' is not assignable to type 'number'
```

Is there an unwrapping happening somewhere? 🤔 
```

    const double = computed<number>(() => count.value * 2);

    // writable, will have a type of Ref<T>
    const triple = computed<number>({
      get: () => count.value * 3,

>       set: (value: number) => value,

Reviewer: the `value:number` is not required, even without `<number>` typescript is able to infer the type correct.

Ideally typescript should be able to infer 80% of the cases, we just need to give to provide some information when those types are not inferred automatically.

Author: I think I will add this to the beginning of the TS section 👍🏻

>     let count = ref<number>(0);

Reviewer: should we remove the `<number>` from this examples? because typescript will be able to infer the correct types in this setup.

# TypeScript Support

> > [Vue CLI](https://cli.vuejs.org) provides built-in TypeScript tooling support.

Reviewer 3: Is this true as of a particular version? If so, we should say so- or even which version is best for TS

Author: It's true since Vue CLI 3.x so we can leave it 🎉

## Official Declaration in NPM Packages

> A static type system can help prevent many potential runtime errors, especially as applications grow. That's why Vue is written in TypeScript. This means you don't need any additional tooling to use TypeScript with Vue - it has a first-class citizen support

Reviewer 3: Second sentence is a fragment, how about:

>A static type system can help prevent many potential runtime errors as applications grow, which is why Vue 3 is written in TypeScript.

Reviewer 3: Also missing a period at the end of the paragraph

Author: Fixed! Thank you @sdras 🙇🏻‍♀️

year.value = 2020 // ok!
```

::: tip Note
If the type of the generic is unknown, it's recommended to cast `ref` to `Ref<T>`.
:::

### Typing `reactive`

When typing a `reactive` property, we can use use interfaces:

```ts
import { defineComponent, reactive } from 'vue'

interface Book {
  title: string
  year?: number
}

export default defineComponent({
  name: 'HelloWorld',
  setup() {
    const book: Book = reactive({ title: 'Vue 3 Guide' })
    // or
    const book = reactive({ title: 'Vue 3 Guide' }) as Book
  }
})
```

Reactive should have the same type as the argument passed to `reactive` method. If there's an unwrapped `ref`, within `reactive` object, the types will differ:

```ts
export default defineComponent({
  name: 'HelloWorld',
  setup() {
    const book: Book = reactive({ title: 'Vue 3 Guide' })

    book.year = ref(2020) // 'Ref<number>' is not assignable to type 'number'
  }
})
```

### Typing `computed`

For computed, we could specify a type with getter only, or with getter and setter:

```ts
import { defineComponent, ref, computed } from "vue";

export default defineComponent({
  name: "HelloWorld",
  setup() {
    let count = ref(0);

> 

Reviewer: can we have an example with `computed` automatically infer the return type? 

The `read-only` bellow can become override computed type

Author: @pikax done!
